### ISSUE TITLE
NA SM: 100% CPU on client disconnection

### ISSUE THREAD

**Issue by ael-code** [mercury-hpc/mercury#229]
### Context
The bug affect the na_sm plugin. When the `HG_UTIL_HAS_SYSEPOLL_H` is defined.
I first open #228 with some hints about the bug, but then I realized which was the root cause of the problem that I will describe in this new issue.

### High level behavior 
I discovered the bug while using [margo](https://xgitlab.cels.anl.gov/sds/margo) with the na_sm plugin. Margo uses a progress thread that is invoking `HG_Progress` continuously with a timeout of `100`.
I noticed that after a client close a communication socket, this progress thread was consuming 100% of the CPU.

### In depth analysis
After a client have been close its communication socket, a call to `epoll_wait` made by the server on that specific socket will return an event of type `EPOLLHUP`. This is a normal behavior described in the man page of `epoll_ctl`.

In mercury, the `EPOLLHUP` event is not handled at all. This means that the socket is neither closed nor de-registered from its epoll set when a na_sm client disconnects.
 
The result is that, after a client have perform a clean shutdown, on the server side, all the subsequent call to `hg_poll_wait()` will always find an event of type `EPOLLHUP` and they will try to process it.
The socket that produced the event is marked as `NA_SM_SOCK_DONE`, thus the `na_sm_progress_sock` function will simply turns into a no-op. And it will return almost immediately `HG_SUCCESS` without making any actual progress.

In this scenario the `hg_core_progress_poll` function will consume 100% of the CPU trying to issue repeatedly the `hg_poll_wait()` function till it reach the given timeout.

Practically we are busy-waiting by using the `epoll_wait` function :)

Except for the 100% CPU consumption, the server is creating a new socket for each new client connection without closing them. This, on a long run, will lead to the following problems:
 - the number of file descriptor is limited and will terminate soon, by making impossible to accept new connections.
 - every new socket is add to the progress poll set, by making it bigger and bigger on every new client connection. The increasing size of the epoll set will make the `epoll_wait` function slower on a long run.

### Quick fix
I've solved the issue by removing the socket from its epoll set after en event of type `EPOLLHUP` is encountered. The bug is actually solved, but now I'm experiencing problems during the shutdown. In fact the socket still remains registered in the `poll_addr_queue` and mercury will try to perform a `na_sm_addr_free` during the server shoutdown.

### Conclusions
In a client-server architecture, it is normal that multiple clients will connect to the server and then disconnect. I think that Mercury should handle this scenario properly. It should take advantage of the `EPOLLHUP` event to close the specific socket and cleanup all the resources related to that client.

**Comment by soumagne** [comment]
Thanks, yes that makes sense now. I will have a look and see how to fix it properly.

**Comment by ael-code** [comment]
Any news on this?

**Comment by soumagne** [comment]
Sorry no not yet, I would actually like to rework the SM plugin a little to completely get rid of the "connection-oriented" code so that issue would also go away at the same time...
